### ISSUE TITLE
ghpc deploy ends up in bad state when instance creation fails due to transient problem

### ISSUE THREAD

**Issue by yaroslavvb** [GoogleCloudPlatform/hpc-toolkit#1536]
TLDR; "ghpc deploy" fails to notice if instance creation fails due to a temporary issue for some of the nodes. Using "ghpc deploy again" doesn't fix the problem.

1. How should cluster be redeployed in such setting?
2. How would I get extra logs or know there's a problem?

For instance, I've launched "ghpc deploy alpha3" using blueprint [here](https://www.dropbox.com/scl/fi/utiz2wmsl38sgi4fwayk1/jun30-alpha.yaml?rlkey=jyloyk5pc6kiwvwqyrd818bid&dl=0
) and it only launched some of the nodes:
```
alpha3-controller                 asia-southeast1-c  c2-standard-4                10.0.0.5      34.16.15.10  RUNNING
alpha3-login-3sa73t1e-001         asia-southeast1-c  n2-standard-8                10.0.0.6                      RUNNING
alpha3-tiny-ghpc-0                asia-southeast1-c  n2-standard-2                10.0.0.2                      RUNNING
alpha3-ultra-ghpc-14              asia-southeast1-c  a2-ultragpu-8g               10.0.0.1                      RUNNING
```

For problem the machines, the node creation command failed with message `VM_MIN_COUNT_NOT_REACHED,TIMEOUT`

```
resourceName: "projects/contextual-research-common/zones/asia-southeast1-c/instances/alpha3-ultra-ghpc-16"
serviceName: "compute.googleapis.com"
status: {
code: 13
message: "VM_MIN_COUNT_NOT_REACHED,TIMEOUT"
```

This is the terraform [directory](https://www.dropbox.com/scl/fi/ppum5bzqsr7oymgy2wkgb/jun30-alpha3.zip?rlkey=3kf82chl65ipl8s0gdjmyh4up&dl=0
) created by ghpc toolkit.

**Comment by mr0re1** [comment]
The error `VM_MIN_COUNT_NOT_REACHED` frequently correlates  with "stockouts", especially given a large number (`22`) of `a2-ultragpu-8g`.
You can find more logs in `/var/log/slurm/resume.log` on controller node (one can SSH from GCP UI).

**Comment by yaroslavvb** [comment]
Thanks for the tip. Do you recommend what to do if this error is transient?

Ie, do ghpc deploy again?

tear down and redeploy cluster causes /opt/apps to be lost and want to
avoid that

**Comment by nick-stroud** [comment]
From the blueprint in the original post it appears your cluster is trying to provision 22 static A2 machines for the GPU cluster. From the information provided it appears that only a single A2 machine was provisioned `alpha3-ultra-ghpc-14`.

If static nodes cannot be provisioned, Slurm will continue to retry to bring up the nodes once a minute. As stated by @mr0re1, it would be helpful to get the logs from `/var/log/slurm/resume.log` as this will record each attempt to bring up the nodes. 

In my attempt to reproduce the issue I see the following error in `/var/log/slurm/resume.log` every minute:

```
2023-07-06 05:21:03,861 INFO: resume nstest01-debug-ghpc-[0-11]
2023-07-06 05:21:05,132 ERROR: bulkInsert API failures: ('nstest01-debug-ghpc:0', <HttpError 503 when requesting https://compute.googleapis.com/compute/v1/projects/hpc-toolkit-dev/regions/us-central1/instances/bulkInsert?alt=json returned "Region does not currently have sufficient capacity for the requested resources.". Details: "[{'message': 'Region does not currently have sufficient capacity for the requested resources.', 'domain': 'global', 'reason': 'insufficientCapacity'}]">)
```

If the issues really is capacity then just waiting should resolve the issue as the nodes will be provisioned once capacity is available, but it is possible that there is some other issue which is why the logs requested are important information. 


I see that in your blueprint you have `enable_reconfigure` set to true. It should not be necessary to destroy the cluster to make configuration changes. For example if you change the value of `node_count_static` and call the following commands the cluster will be updated while staying live.

```
./ghpc create -w <blueprint name>  # recreate the deployment folder
./ghpc deploy <deployment folder>  # apply changes 
```

**Comment by nick-stroud** [comment]
It is not clear why the nodes are failing to start in the first place. The `VM_MIN_COUNT_NOT_REACHED` is an error coming from GCP and not from Slurm or the HPC Toolkit. I suggest following up with GCP support to root cause that issue. 

I will add here some useful actions that can help the cluster to recover if this happens.

If it appears that the node was never created and logs found in /var/log/slurm/* do not indicate that Slurm is taking action to create the node, you can prompt slurm to start the node by calling the following from the controller:

```
sudo /slurm/scripts/resume.py <name of missing node>
```

This will tell the controller to try to create the node.

If the node appears to be up in the google cloud UI but is showing as down by the controller. You can SSH to the down compute node and call the following command:

```
cd / && sudo /slurm/scripts/setup.py
```

This should allow the node to rejoin the cluster.
